controller/src/scheduler.rs

```diff
@@ -0,0 +1,160 @@
+const SCHEDULER_CRON_EXPRESSION_ENV_VAR: &str = "SCHEDULER_CRON_EXPRESSION";
+
+/// The module-wide result type.
+type Result<T> = std::result::Result<T, scheduler_error::Error>;
+
+pub struct BrupopCronScheduler {
+    scheduler_cron_expression: String,
+}
+
+impl BrupopCronScheduler {
+    pub fn from_environment() -> Result<Self> {
+        let scheduler_cron_expression = env::var(SCHEDULER_CRON_EXPRESSION_ENV_VAR).context(
+            scheduler_error::MissingEnvVariableSnafu {
+                variable: SCHEDULER_CRON_EXPRESSION_ENV_VAR.to_string(),
+            },
+        )?;
+        Ok(BrupopCronScheduler {
+            scheduler_cron_expression: scheduler_cron_expression,
+        })
+    }
+
+    /// Cron expression parser and schedule explorer
+    pub async fn generate_schedule(&self) -> Result<Schedule> {
+        let schedule = Schedule::from_str(&self.scheduler_cron_expression).context(
+            scheduler_error::GenerateScheduleFailedSnafu {
+                cron: &self.scheduler_cron_expression,
+            },
+        )?;
+
+        Ok(schedule)
+    }
+}
+
+// Cron expression can be configured to a time window or a specific trigger time.
+// specific trigger time: 0 0 10 * * Mon */Every Monday at 10AM.
+// maintenance window (time window): * * 10-12 * * MON */Every Monday between 10:00 and 12:00.
+
+// brupop controller needs to use different logics to deal with specific trigger time or
+// maintenance window (time window)
+// => specific trigger time: trigger brupop update and complete all waitingForUpdate nodes.
+// => maintenance window (time window): trigger brupop update within time window. If current
+// time isn't within the time window, controller shouldn't have any action on it.
+pub enum MaintenanceTime {
+    Windows(cron::Schedule),
+    Oneshot(cron::Schedule),
+}
+
+impl MaintenanceTime {
+    pub fn new(cron: cron::Schedule) -> Result<Self> {
+        if calculate_duration_between_scheduled_datetimes(&cron)?.num_seconds() == 1 {
+            Ok(MaintenanceTime::Windows(cron))
+        } else {
+            Ok(MaintenanceTime::Oneshot(cron))
+        }
+    }
+
+    /// Make controller sleep if no more active nodes and current is outside of a scheduled window.
+    pub async fn prepare_and_trigger_updates(&self) -> Result<()> {
+        // calculate the duration between current time and next scheduled time, and then make
+        // controller sleep until next scheduled time point.
+        match self {
+            MaintenanceTime::Windows(s) => sleep_until_next_scheduled_time(s).await,
+            MaintenanceTime::Oneshot(s) => sleep_until_next_scheduled_time(s).await,
+        }
+    }
+
+    /// Determine when controller needs discontinue updates.
+    /// specific trigger time => never discontinue updates.
+    /// maintenance window (time window): discontinue updates when current is outside of a scheduled window.
+    pub fn discontinue_updates(&self) -> bool {
```

Comment:
nit: `discontinue_updates` sounds like an imperative command, I'd rename this to `should_discontinue_updates`.